### ISSUE TITLE
Magento 2.2.0 does not support nullable return types on PHP 7.0.24

### ISSUE THREAD

**Issue by esimov** [magento/magento2#11264]
### Preconditions

1. Magento 2.2.0
2. PHP 7.0.24

### Steps to reproduce

Checking the system requirements for Magento 2.2.0 on http://devdocs.magento.com/guides/v2.2/install-gde/system-requirements-tech.html#php it shows that PHP 7.0 is supported, however the VarTag.php file under vendor/zendframework/zend-code/src/Reflection/DocBlock/Tag/ contains a few methods which returns nullable return types. Like the methods below:
```php
public function getVariableName() : ?string
{
    return $this->variableName;
}

public function getDescription() : ?string
{
    return $this->description;
}
```

The nullable return types are not yet supported on PHP7.0. They were introduced from PHP7.1 on. This means that the actions using the methods above (like the login action) thrown syntax error. Please correct the system requirements.

**Comment by orlangur** [comment]
Strange that it wasn't revealed by integration tests. Is this code actually used in Magento 2? Could you provide some scenario?

Looks like `3.1.0` release of `Zend_Code` didn't support PHP 7.1: https://github.com/zendframework/zend-code/blob/release-3.1.0/composer.json#L16
While `3.2.0` release does not support PHP 7.0: https://github.com/zendframework/zend-code/blob/release-3.2.0/composer.json#L16

Thus there are only two options for Magento 2 to solve the problem directly:
- drop PHP 7.0 support
- maintain separate composer files for PHP 7.0 and PHP 7.1

Indirect solution is to fully cover all code which relies on `Zend_Code` with integration tests so that PHP 7.0 workability is preserved.

**Comment by orlangur** [comment]
According to @ytorbyk we can observe problem with following scenario:

1. Use PHP 7.1
2. Create new Magento project via Composer
3. Try performing DI compilation
4. Exception observed due to Magento compiler not supporting nullable `?string` notation

If we compile using PHP 7.0, it will work just fine under PHP 7.1 then.

**Comment by orlangur** [comment]
Just found out Magento is using `3.1.0` version of `zend-code`: https://github.com/magento/magento2/blob/2.2/composer.lock#L2022

So, there is no problem with PHP 7.0 compatibility, just that this component does not officially support PHP 7.1 but maybe it does not affect Magento.

Those who somehow installed `3.2.0` version of `zend-code` seems to be in trouble though.

**Comment by esimov** [comment]
Then this means there is an error on zend-code `composer.json`, because officially does not support PHP 7.1, however they are using a PHP feature which exists only on PHP 7.1.

**Comment by orlangur** [comment]
Nope, `zend-code` is fine: nullable types are only used in `3.2.0` release I believe (didn't check though).

**Comment by mzeis** [comment]
We have had this problem before. Depending on the PHP version of your system, a `composer update` might install different packages if `composer.json` doesn't define a target platform (= "main" PHP version). See [platform](https://getcomposer.org/doc/06-config.md#platform). It might be a good idea to add this to Magento's `composer.json`.

**Comment by fooman** [comment]
@esimov do you mind confirming how you installed Magento 2.2.0

**Comment by orlangur** [comment]
> See platform. It might be a good idea to add this to Magento's `composer.json`.

No, no, no, hell no :) Just like `--ignore-platform-reqs` it's just a way of faking and get some headache. We need to have a list of really compatible packages for every PHP version.

**Comment by mzeis** [comment]
How do you release one version Magento 2.x.y for different PHP versions? Or make sure that every dev uses the correct PHP version for updating packages? :)

**Comment by orlangur** [comment]
It is already tested on CICD. If some combinations are missing in matrix yet, they should be added :)

That's basically exactly what Composer is intended for - produce list of packages compatible with each other and the platform.

@fooman could you please try `composer create-project --repository-url="https://repo.magento.com/" magento/project-community-edition m2zendcode` under PHP 7.1? I don't have it in my hand right now, according to @ytorbyk it installs `zend-code` version `3.2.0` and `setup:di:compile` is broken.

**Comment by esimov** [comment]
@fooman it was an upgrade from Magento 2.1.8 via `composer.json`, so not with `composer require`.

**Comment by fooman** [comment]
@mzeis agree with @orlangur Magento itself should not add platform to the distributed `composer.json`. However any developer/agency that is installing and working in a mixed php version environment should add platform to their _own_ copy of `composer.json` to avoid exactly this type of issue.

The issue that we do have is that Magento pre-generates a zip download available here
https://magento.com/tech-resources/download for Magento 2.2 that makes a determination on behalf of the user and generates under php 7.0. (the zip comes with zend/code 3.1.0)

What Magento should do (if they want to continue offering the zip download) is, at a minimum, use platform 7.0 and 7.1 to create two separate zip downloads (I'll create a separate issue for this).

@esimov can you confirm if you do have a mixed environment of php versions or if you are already using platform or ignore platform reqs? I did an upgrade from 2.1.9 to 2.2.0 via composer on php 7.0 and ended up with the right version:

```
$ composer show zendframework/zend-code
name     : zendframework/zend-code
descrip. : provides facilities to generate arbitrary code using an object oriented interface
keywords : code, zf2
versions : * 3.1.0
```

@orlangur also tested the following on 7.1
```
$ composer create-project --repository-url=https://repo.magento.com/ magento/project-community-edition m2
$ cd m2
$ bin/magento install _____
...
$ composer show zendframework/zend-code
name     : zendframework/zend-code
descrip. : provides facilities to generate arbitrary code using an object oriented interface
keywords : code, zf2
versions : * 3.2.0
$ bin/magento setup:di:compile
Compilation was started.
Interception cache generation... 7/7 [============================] 100% 17 secs 244.0 MiB
Generated code and dependency injection configuration successfully.
```
 seems something is missing from @ytorbyk steps.

**Comment by orlangur** [comment]
@fooman, thanks for testing! @ytorbyk confirmed he cannot reproduce this issue anymore so I suppose the problem was with mixed PHP environment as well.

Waiting for feedback from @esimov.

**Comment by esimov** [comment]
@orlangur @fooman yes, it happened on a mixed PHP environment, but with PHP 7.1 this issue has gone. Once i switched to PHP 7.1 the `composer update` ran successfully, without to ignore platform reqs, but the syntax error for nullable types thrown by PHP 7.0 however was not the case for PHP 7.1, because the lately however supports the nullable types.

**Comment by orlangur** [comment]
@esimov thanks! Looks like Composer does it job well and `zend-code` is installed with version compatible with platform you use for both PHP 7.0 and 7.1.

However, if you do `composer update` under PHP 7.1 and then try to use same package versions under 7.0, it will lead to troubles but it is quite expectable.

**Comment by scholtz** [comment]
Just to copy from the other thread, i believe that if you run PHP7.1 CLI version for composer, and run PHP 7.0 on PHP-FPM, the Magento should warn administrator that something is broken, because users may get unexpected blank screens on unknown places (everywhere where php 7.1 specifi code is used) when php get parse error..

I also agree that if you install using cli 7.1, fpm should be 7.1+, but if webmaster do not notice or forget to update php on webserver, he should get notified somehow..

**Comment by orlangur** [comment]
> the Magento should warn administrator that something is broken

> if webmaster do not notice or forget to update php on webserver, he should get notified somehow..

Nope, just that Composer should be used properly. Whoever face with this issue will notice parse error immediately and will perform necessary actions to fix mistake.

**Comment by orlangur** [comment]
Hi @scholtz, I locked conversation here just to stop flame, feel free to comment if you have something constructive to add. Just make sure that your comments comply with [Code of Conduct](https://github.com/magento/magento2/wiki/Magento-Code-of-Conduct).

FYI, this issue was accidentally fixed in `2.2-develop` already, zend-code 3.1.x is currently used for both PHP 7.0 and 7.1, the downside is that it has less features compared to zend-code 3.3.x.
